File: dart/anagram/lib/anagram.dart
class Anagram {
  List<String> findAnagrams(String subject, List<String> candidates) =>
      candidates
          .map((word) => MapEntry(word, word.toLowerCase()))
          .where((word) => word.value != subject.toLowerCase())
          .where((word) => word.value.isAnagramOf(subject.toLowerCase()))
          .map((word) => word.key)
          .toList();
}

extension AnagramUtils on String {
  bool isAnagramOf(String subject) => sort() == subject.sort();

  String sort() => (split("")..sort()).join();
}

File: kotlin/anagram/src/main/kotlin/Anagram.kt
class Anagram(subject: String) {

    private val subject: String = subject.decapitalize()

    fun match(candidates: List<String>): Set<String> =
        candidates
            .filter { candidate -> subject isAnagramOf candidate }
            .toSet()

    private infix fun String.isAnagramOf(candidate: String): Boolean =
        groupBy { it } == candidate.decapitalize().groupBy { it }
}

